### PR TITLE
JsonSchema Support

### PR DESCRIPTION
**Pull request by mlodi-hqs**
Adds support for `JsonSchema` to qoqo and roqoqo.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ This changelog track changes to the qoqo project starting at version v0.5.0
 ## 1.6.0-alpha.1
 
 * Renamed the PNRDetection operation to PhotonDetection.
+* Added support for JsonSchema.
```
Comment (kbarkhqs) [inline CHANGELOG.md:8]: You'll need to put this in a 1.6.0-alpha.2 section and update to 1.6.0-alpha.2, as version 1.6.0-alpha.1 has already been released

File: qoqo/Cargo.toml [hunk 2]
```diff
@@ -55,4 +56,5 @@ no-default-features = true
 extension-module = ["pyo3/extension-module", "circuitdag"]
 default = ["extension-module"]
```
Comment (kbarkhqs) [inline qoqo/Cargo.toml:57]: Should the json_schema feature be added to the defaults?

File: qoqo/tests/integration/circuit.rs [hunk 3]
```diff
@@ -412,6 +413,20 @@ fn test_to_from_json() {
     })
 }
 
+/// Test json_schema function of Circuit
+#[cfg(feature = "json_schema")]
+#[test]
+fn test_json_schema() {
+    pyo3::prepare_freethreaded_python();
+    pyo3::Python::with_gil(|py| {
+        let circuit = new_circuit(py);
```
Comment (kbarkhqs) [inline qoqo/tests/integration/circuit.rs:422]: I'd recommend adding an operation to the circuit for additional testing, but that may just be personal preference

File: qoqo/tests/integration/measurements/mod.rs [hunk 4]
```diff
@@ -10,6 +10,8 @@
 // express or implied. See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![allow(clippy::redundant_clone)]
```
Comment (kbarkhqs) [inline qoqo/tests/integration/measurements/mod.rs:13]: Why has this been added?

File: qoqo/tests/integration/operations/pragma_operations.rs [hunk 5]
```diff
@@ -2585,3 +2585,106 @@ fn test_pyo3_remapqubits_error(input_operation: Operation) {
...
+#[test_case(PragmaOperation::from(PragmaDepolarising::new(0, CalculatorFloat::from(0.005), CalculatorFloat::from(0.02))); "PragmaDepolarising")]
+#[test_case(PragmaOperation::from(PragmaDephasing::new(0, CalculatorFloat::from(0.005), CalculatorFloat::from(0.02))); "PragmaDephasing")]
+#[test_case(PragmaOperation::from(PragmaRandomNoise::new(0, CalculatorFloat::from(0.005), CalculatorFloat::from(0.02), CalculatorFloat::from(0.01))); "PragmaRandomNoise")]
+#[test_case(PragmaOperation::from(PragmaGeneralNoise::new(0, CalculatorFloat::from(0.005),  operators())); "PragmaGeneralNoise")]
+#[test_case(PragmaOperation::from(PragmaConditional::new(String::from("ro"), 1, create_circuit())); "PragmaConditional")]
+#[test_case(PragmaOperation::from(PragmaControlledCircuit::new( 1, create_circuit())); "PragmaControlledCircuit")]
+#[test_case(PragmaOperation::from(PragmaLoop::new(CalculatorFloat::from("number_t"), Circuit::default())); "PragmaLoop")]
+fn test_pyo3_json_schema(operation: PragmaOperation) {
+    let rust_schema = match operation {
+        PragmaOperation::PragmaSetNumberOfMeasurements(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/pragma_operations.rs:2611]: This one isn't in the list above

File: qoqo/tests/integration/operations/pragma_operations.rs [hunk 6]
```diff
@@ -2585,3 +2585,106 @@ fn test_pyo3_remapqubits_error(input_operation: Operation) {
...
+        PragmaOperation::PragmaSetStateVector(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaSetStateVector)).unwrap()
+        }
+        PragmaOperation::PragmaSetDensityMatrix(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaSetDensityMatrix)).unwrap()
+        }
+        PragmaOperation::PragmaRepeatGate(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaRepeatGate)).unwrap()
+        }
+        PragmaOperation::PragmaOverrotation(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/pragma_operations.rs:2624]: This one is not in the list above

File: qoqo/tests/integration/operations/pragma_operations.rs [hunk 7]
```diff
@@ -2585,3 +2585,106 @@ fn test_pyo3_remapqubits_error(input_operation: Operation) {
...
+        PragmaOperation::PragmaRandomNoise(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaRandomNoise)).unwrap()
+        }
+        PragmaOperation::PragmaGeneralNoise(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaGeneralNoise)).unwrap()
+        }
+        PragmaOperation::PragmaConditional(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PragmaConditional)).unwrap()
+        }
+        PragmaOperation::PragmaGetPauliProduct(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/pragma_operations.rs:2668]: This one is not in the list above

File: qoqo/tests/integration/operations/single_qubit_gate_operations.rs [hunk 8]
```diff
@@ -2586,3 +2586,110 @@ fn test_pyo3_richcmp(definition_1: Operation, definition_2: Operation) {
...
+#[test_case(SingleQubitGateOperation::from(PhaseShiftState0::new(0, CalculatorFloat::from(0.0))); "PhaseShiftState0")]
+#[test_case(SingleQubitGateOperation::from(PhaseShiftState1::new(0, CalculatorFloat::from(0.0))); "PhaseShiftState1")]
+#[test_case(SingleQubitGateOperation::from(GPi::new(0, CalculatorFloat::from(0.0))); "GPi")]
+#[test_case(SingleQubitGateOperation::from(GPi2::new(0, CalculatorFloat::from(0.0))); "GPi2")]
+fn test_pyo3_json_schema(operation: SingleQubitGateOperation) {
+    let rust_schema = match operation {
+        SingleQubitGateOperation::SingleQubitGate(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(SingleQubitGate)).unwrap()
+        }
+        SingleQubitGateOperation::RotateZ(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/single_qubit_gate_operations.rs:2631]: Missing above

File: qoqo/tests/integration/operations/single_qubit_gate_operations.rs [hunk 9]
```diff
@@ -2586,3 +2586,110 @@ fn test_pyo3_richcmp(definition_1: Operation, definition_2: Operation) {
...
+        SingleQubitGateOperation::TGate(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(TGate)).unwrap()
+        }
+        SingleQubitGateOperation::PhaseShiftState1(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PhaseShiftState1)).unwrap()
+        }
+        SingleQubitGateOperation::PhaseShiftState0(_) => {
+            serde_json::to_string_pretty(&schemars::schema_for!(PhaseShiftState0)).unwrap()
+        }
+        SingleQubitGateOperation::RotateAroundSphericalAxis(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/single_qubit_gate_operations.rs:2670]: Missing above

File: qoqo/tests/integration/operations/single_qubit_gate_operations.rs [hunk 10]
```diff
@@ -2586,3 +2586,110 @@ fn test_pyo3_richcmp(definition_1: Operation, definition_2: Operation) {
...
+            serde_json::to_string_pretty(&schemars::schema_for!(RotateAroundSphericalAxis)).unwrap()
+        }
+        SingleQubitGateOperation::RotateXY(_) => {
```
Comment (kbarkhqs) [inline qoqo/tests/integration/operations/single_qubit_gate_operations.rs:2673]: Missing above

File: roqoqo/tests/integration/operations/mod.rs [hunk 11]
```diff
@@ -10,6 +10,8 @@
 // express or implied. See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![allow(clippy::redundant_clone)]
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/mod.rs:13]: Should this be added here?

File: roqoqo/tests/integration/operations/multi_qubit_gate_operations.rs [hunk 12]
```diff
@@ -565,3 +569,33 @@ fn test_rotatex_powercf_multi_qubit_zz(theta: CalculatorFloat, power: Calculator
     assert_eq!(power_gate, test_gate);
     assert_eq!(power_gate.theta(), test_gate.theta());
 }
+
+/// Test JsonSchema trait
+#[cfg(feature = "json_schema")]
+#[test_case(MultiQubitGateOperation::from(MultiQubitZZ::new(vec![0, 1, 2, 3], 0.23.into())); "MultiQubitZZ")]
+#[test_case(MultiQubitGateOperation::from(MultiQubitMS::new(vec![0, 1, 2], 0.45.into())); "MultiQubitMS")]
+pub fn test_json_schema_multi_qubit_gate_operations(gate: MultiQubitGateOperation) {
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/multi_qubit_gate_operations.rs:577]: Would this kind of set-up be better for the define operations, bosonic operations and measurement operations?

File: roqoqo/tests/integration/operations/single_qubit_gate_operations.rs [hunk 13]
```diff
@@ -2350,3 +2354,86 @@ pub fn test_tags(gate: SingleQubitGateOperation, tags: Vec<&str>) {
         assert_eq!(gate.tags()[i], tags[i]);
     }
 }
+
+/// Test JsonSchema trait
+#[cfg(feature = "json_schema")]
+#[test_case(SingleQubitGateOperation::from(RotateX::new(0, CalculatorFloat::from("theta"))); "RotateX")]
+#[test_case(SingleQubitGateOperation::from(RotateY::new(0, CalculatorFloat::from("theta"))); "RotateY")]
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/single_qubit_gate_operations.rs:2361]: RotateXY, RotateZ, and SingleQubitGate are missing

File: roqoqo/tests/integration/measurements/mod.rs [hunk 14]
```diff
@@ -10,6 +10,8 @@
 // express or implied. See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![allow(clippy::redundant_clone)]
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/measurements/mod.rs:13]: Why has this been added?

**Review by kbarkhqs — CHANGES_REQUESTED** [review]
Some tests missing and some questions sent over

File: qoqo/tests/integration/measurements/mod.rs (hunk 4 above)
Comment (mlodi-hqs) [reply qoqo/tests/integration/measurements/mod.rs:13]: it suppresses clippy warnings in the test regarding duplicate clones, they are generated when `.clone()` is tested. 

It's just a preference of mine but I can revert it, it's not necessary of course.

File: roqoqo/tests/integration/operations/mod.rs (hunk 11 above)
Comment (mlodi-hqs) [reply roqoqo/tests/integration/operations/mod.rs:13]: Same as above

File: roqoqo/tests/integration/operations/multi_qubit_gate_operations.rs (hunk 12 above)
Comment (mlodi-hqs) [reply roqoqo/tests/integration/operations/multi_qubit_gate_operations.rs:577]: I generally followed the structure that was already there in the tests. But I've just realized that for the MultiQubitGateOperation enum I didn't.

File: roqoqo/tests/integration/measurements/mod.rs (hunk 14 above)
Comment (mlodi-hqs) [reply roqoqo/tests/integration/measurements/mod.rs:13]: Same as above

File: qoqo/tests/integration/measurements/mod.rs (hunk 4 above)
Comment (mlodi-hqs) [reply qoqo/tests/integration/measurements/mod.rs]: Fixed with new solution.

**Review by kbarkhqs — APPROVED** [review]

**Pull request merged**